store/bolt/user_store.go

```diff
@@ -0,0 +1,124 @@
+package bolt
+
+import (
+	"errors"
+
+	"github.com/boltdb/bolt"
+	"github.com/gopheracademy/congo"
+	"github.com/gopheracademy/congo/store/bolt/internal"
+)
+
+var (
+	// ErrUserNotFound is returned when mutating a user that doesn't exist.
+	ErrUserNotFound = errors.New("user not found")
+
+	// ErrEmailRequired is returned when creating a user without an email.
+	ErrEmailRequired = errors.New("email address required")
+)
+
+// UserStore represents the data store for user data.
+type UserStore struct {
+	db *bolt.DB
+}
+
+// initialize creates the top-level buckets for the store.
+func (s *UserStore) init() error {
+	return s.db.Update(func(tx *bolt.Tx) error {
+		tx.CreateBucketIfNotExists([]byte("users"))
+		tx.CreateBucketIfNotExists([]byte("users.email"))
+		return nil
+	})
+}
+
+// User returns a user by ID.
+func (s *UserStore) User(id uint64) (u *congo.User, err error) {
+	err = s.db.View(func(tx *bolt.Tx) error {
+		if v := tx.Bucket([]byte("users")).Get(u64tob(id)); v != nil {
+			u, err = internal.UnmarshalUser(v)
+			return err
+		}
+		return nil
+	})
+	return
+}
+
+// UserByEmail returns a user by email address.
+func (s *UserStore) UserByEmail(email string) (u *congo.User, err error) {
+	err = s.db.View(func(tx *bolt.Tx) error {
+		// Lookup user ID from index.
+		key := tx.Bucket([]byte("users.email")).Get([]byte(email))
+		if key == nil {
+			return nil
+		}
+
+		// Retrieve the user by ID.
+		if v := tx.Bucket([]byte("users")).Get(key); v != nil {
+			u, err = internal.UnmarshalUser(v)
+			return err
+		}
+		return nil
+	})
+	return
+}
```

Comment:
Bolt has no concept of indexes so we have to build our own. This function works by looking up the ID of a user by their email address which has been saved to the `users.email` bucket in `CreateUser()`. If an ID was found then it's looked up on L55 and unmarshaled and returned.